src/compas_timber/ghpython/workflow.py

```diff
@@ -35,13 +35,37 @@ def comply(self, beams):
         raise NotImplementedError
 
 
+class DirectRule(JointRule):
+    """for a given connection topology type (L,T,X,I,K...), this rule assigns a joint type."""
+
+    def __init__(self, joint_type, beams, **kwargs):
+        self.beams = beams
+        self.joint_type = joint_type
+        self.kwargs = kwargs
+
+    def ToString(self):
+        # GH doesn't know
+        return repr(self)
+
+    def __repr__(self):
+        return "{}({}, {})".format(DirectRule, self.beams, self.joint_type)
+
+    def comply(self, beams):
+        try:
+            return set(self.beams) == set(beams)
+        except KeyError:
```

Comment:
When is a `KeyError` raised?

Reply:
Hmm good point. what kind of error should we raise? maybe comply should check for topo match...?